File: 3133_minimum_array_end.java
class Solution {
    public long minEnd(int n, int x) {
        long res = 0L;
        int m = 0, p = 0;
        for (long u = x, v = n - 1; m < 56; m++) {
            res |= ((u & (1L << m)) != 0 ? 1L : (v >> p++) & 1L) << m;
        }
        return res;
    }
}

File: 3133_minimum_array_end.cpp
class Solution {
public:
    long long minEnd(int n, int x) {
        bitset<64> p(x), q(n - 1), r;
        for (int s = 0, t = 0; s < 56; s++) {
            r[s] = (p[s]) ? 1 : q[t++];
        }
        return r.to_ullong();
    }
};
